publish-mqtt/src/main.rs

```diff
@@ -384,88 +399,77 @@ async fn check_for_sensors(
     Ok(())
 }
 
-async fn connect_sensor_at_idx(
+async fn connect_sensor_with_id(
     state: Arc<Mutex<SensorState>>,
     session: &MijiaSession,
-    idx: usize,
+    id: DeviceId,
 ) -> Result<(), anyhow::Error> {
-    // Get a copy of the sensor, and update the state of the original to `Connecting`.
-    let mut sensor = {
+    // Update the state of the sensor to `Connecting`.
+    {
         let mut state = state.lock().await;
-        let sensor = state.sensors[idx].clone();
-        state.sensors[idx].connection_status = ConnectionStatus::Connecting {
+        let sensor = state.sensors.get_mut_by_id(&id).unwrap();
+        println!(
+            "Trying to connect to {} from status: {:?}",
+            sensor.name, sensor.connection_status
+        );
+        sensor.connection_status = ConnectionStatus::Connecting {
             reserved_until: Instant::now() + SENSOR_CONNECT_RESERVATION_TIMEOUT,
         };
-        sensor
     };
-    // Try to connect to a sensor.
-    println!("Trying to connect to {}", sensor.name);
-    let status = connect_start_sensor(session, &mut sensor).await;
+    let result = connect_and_subscribe_sensor_or_disconnect(session, &id).await;
+
     let mut state = state.lock().await;
-    match status {
-        Err(e) => {
-            println!(
-                "Failed to connect to {} (now {:?}): {:?}",
-                sensor.name, sensor.connection_status, e
-            );
-        }
+    let SensorState { sensors, homie, .. } = state.deref_mut();
```

Comment:
This destructuring thing isn't actually needed, you can just use the `let state = &mut *state.lock().await;" trick that we use elsewhere.